Ignore `-C strip` on MSVC

icedrocket opened the pull request:
tl;dr - Define `-Cstrip` to only ever affect the binary; no other build artifacts.

This is necessary to improve cross-platform behavior consistency: if someone wanted debug information to be contained only in separate files on all platforms, they would set `-Cstrip=symbols` and `-Csplit-debuginfo=packed`, but this would result in no PDB files on MSVC.

Resolves #114215

icedrocket commented:
There may be cases where someone doesn't want a PDB file to be created. One way around this is to prevent PDB files from being created when `split-debuginfo` is `none`.

b-naber commented:
> One way around this is to prevent a PDB file from being created when split-debuginfo is none

I don't understand what you're trying to say here, do you mean that you could prevent a PDB file from being created on MSVC by using `split-debuginfo=off`? `off` is not supported on MSVC, so there wouldn't be a way to prevent a PDB file from being created if we ignore strip afaict.

icedrocket commented:
Currently, the value of the `strip` option determines whether or not a PDB file is created, and I'm proposing to remove that condition or use a different one. The title of the pull request is just my main proposal.

Yes, MSVC doesn't actually support `split-debuginfo=none`, but we can use it as a way to prevent the PDB file from being created. I think the `strip` option should only be applied to the executable itself, not to a separate debug file, as it would break cross-platform behavior consistency.

b-naber commented:
I think it would make sense to treat `split-debuginfo` that way on msvc. That would likely require an mcp though and would also be a prerequisite for landing this pr.

apiraino commented:
earmarking this PR, waiting on that MCP seconding

@rustbot author s-waiting-on-mcp

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6779949760/job/18427842545) [(plain)](https://github.com/rust-lang/rust/commit/cee70c9f91f01856ed608679a457ad5c7b6023d0/checks/18427842545/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=icedrocket
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_2829779d-931a-457a-98c0-8b9be2585351
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=b5fe6d5cc7a9f644abc7682d7373d9a63d988cf8
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_2829779d-931a-457a-98c0-8b9be2585351
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_2829779d-931a-457a-98c0-8b9be2585351
GITHUB_TRIGGERING_ACTOR=icedrocket
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/115120/merge
GITHUB_WORKFLOW_SHA=b5fe6d5cc7a9f644abc7682d7373d9a63d988cf8
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_19_X64=/opt/hostedtoolcache/go/1.19.13/x64
```

</details>

icedrocket commented:
With the MCP considered approved, I made some changes to the behavior of `strip` and `split-debuginfo` on MSVC. I've updated all the relevant code and documentation I could find, but I'd like someone familiar with the compiler code to verify. Some comments in `config.example.toml` have been changed, but no defaults have been changed.

icedrocket commented:
With the MCP considered approved, I made some changes to the behavior of `strip` and `split-debuginfo` on MSVC. I've updated all the relevant code and documentation I could find. The updated documentation states that the behavior has changed since 1.75, which hasn't been released yet, so I'm not sure if that's okay. Also, I'd like someone to re-run the CI that failed, probably because of the request limit.

icedrocket commented:
r? @bjorn3

bors commented:
:umbrella: The latest upstream changes (presumably #117716) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

icedrocket commented:
@rustbot review

bjorn3 commented on compiler/rustc_codegen_ssa/src/back/linker.rs:
```diff
@@ -914,9 +914,12 @@ impl<'a> Linker for MsvcLinker<'a> {
+    fn debuginfo(&mut self, _strip: Strip, natvis_debugger_visualizers: &[PathBuf]) {
+        match self.sess.split_debuginfo() {
+            SplitDebuginfo::Off => {
+                self.cmd.arg("/DEBUG:NONE");
```
`-Csplit-debuginfo=off` is supposed to put all debuginfo in the executable itself. This isn't possible for MSVC, right? And it is definitively not what `/DEBUG:NONE` does. `-Csplit-debuginfo=unpacked` is supposed to keep all debuginfo in their original object files, which `/DEBUG:FASTLINK` does AFAIU.

bjorn3 replied:
> `-Csplit-debuginfo=off` is supposed to put all debuginfo in the executable itself.

In other words, if you use `-Csplit-debuginfo=off` only the executable itself is necessary for debugging. With `-Csplit-debuginfo=packed`, the executable and a single separate file containing the debuginfo is necessary. And with `-Csplit-debuginfo=unpacked` all build artifacts may be necessary for debugging.

icedrocket replied:
The codegen options documentation already states that `split-debuginfo=off` is supported by all platforms except MSVC. Doesn't this imply that if the platform doesn't support including debug information in the executable, it won't generate any debug information?

bjorn3 replied:
It shouldn't have been added to macOS either for the same reason as it shouldn't be added to MSVC.

icedrocket replied:
I was wondering if I would need another MCP and approval to add `split-debuginfo=unpacked` as `/DEBUG:FASTLINK`.

icedrocket replied:
I am wondering if I need another MCP and approval to add `split-debuginfo=unpacked` as `/DEBUG:FASTLINK`.

icedrocket replied:
What do you think about not generating PDB files if `debuginfo=none` instead? AFAIK, DWARF objects are omitted when `debuginfo=none`.

icedrocket replied:
What do you think about not generating PDB files if `debuginfo=none` instead? AFAIU, DWARF objects are omitted when `debuginfo=none`.

icedrocket commented:
I am wondering if I need another MCP and approval to add `split-debuginfo=unpacked` as `/DEBUG:FASTLINK`.

icedrocket commented:
I made some changes based on what the reviewer pointed out.

Instead of changing `-Csplit-debuginfo=off` on MSVC to determine whether to generate a PDB file, I added a fallback behavior for `-Csplit-debuginfo=off` on platforms that do not support storing debug information in the executable.

The fallback behavior would attempt to store the debug information in the separate file, in which case the `strip` option would be applied to the separate file. I changed `strip` option to normally only apply to the executable, not the separate file.

As a result, if both `-Csplit-debuginfo=off` and `-Cstrip=symbols` are set, a PDB file will not be created, in which case Unix ELF will also not include debug information.

icedrocket commented:
I removed the phrase "fallback behavior" and changed the behavior of some options only for MSVC because I felt it was a bit ambiguous. Nothing has changed in the actual implementation.

icedrocket commented:
I'm concerned not only about cross-platform consistency, but also about the possibility that MSVC or MSVC-like compilers may support including debug information in the executable at some point in the future. If that happens, the definition of `strip` will need to change.

Since `-Csplit-debuginfo=off` and `-Cstrip=symbols` is a configuration that completely removes debug information on platforms that use ELF, wouldn't it be better if the same configuration had the same effect on all other platforms?

icedrocket commented:
I'm not only concerned about cross-platform consistency but also about the potential scenario where MSVC or MSVC-like compilers might support including debug information in the executable in the future. If such support is added, the definition of strip will need to be updated.

Given that -Csplit-debuginfo=off and -Cstrip=symbols is a configuration that completely removes debug information on platforms using ELF, wouldn't it be more beneficial if this same configuration had a consistent effect on all other platforms?

icedrocket commented:
I'm not only concerned about cross-platform consistency but also about the potential scenario where MSVC or MSVC-like compilers might support including debug information in the executable in the future. If such support is added, the definition of `strip` will need to be updated.

Given that `-Csplit-debuginfo=off` and `-Cstrip=symbols` is a configuration that completely removes debug information on platforms using ELF, wouldn't it be more beneficial if this same configuration had a consistent effect on all other platforms?

icedrocket commented:
I'm not only concerned about cross-platform consistency but also about the potential scenario where MSVC or MSVC-like compilers might support including debug information in the executable in the future. If such support is added, the definition of `strip` will need to be changed.

Given that `-Csplit-debuginfo=off` and `-Cstrip=symbols` is a configuration that completely removes debug information on platforms using ELF, wouldn't it be more beneficial if this same configuration had a consistent effect on all other platforms?

bjorn3 commented:
> Given that -Csplit-debuginfo=off and -Cstrip=symbols is a configuration that completely removes debug information on platforms using ELF, wouldn't it be more beneficial if this same configuration had a consistent effect on all other platforms?

I think so.

icedrocket commented:
r? compiler-team

petrochenkov commented:
r? @davidtwco 
(This is about split debuginfo.)

davidtwco commented:
I'm not sure about this change, but I appreciate that the behaviour in #114215 isn't desirable.

Intuitively, I'd expect the following behaviour:

- Debuginfo generation should be controlled by `-g`/`-Cdebuginfo`.
- `-Csplit-debuginfo` should only control where debuginfo goes, and its options should only work if supported by the platform.
  - i.e. only in `pdb` files and only `-Csplit-debuginfo=packed` on MSVC
- `-Cstrip` should strip debuginfo/symbols from the binary (not from auxiliary files - though maybe it should, I don't have strong feelings about that).

I think the confusion here is because we decided at some point that `-Cstrip` should influence pdb file generation. On MSVC, if you ask for debuginfo, we should produce a `.pdb`. You can't select anything other than `-Csplit-debuginfo=packed`, because we can't put debuginfo anywhere else on MSVC. If you ask for `-Cstrip=symbols` then that shouldn't affect the generation of a `pdb` file. #114215 would be resolved because we would still have a `pdb` file given that set of options, which is exactly what I would expect.

> `off` is not supported on MSVC, so there wouldn't be a way to prevent a PDB file from being created if we ignore strip afaict.

Reading this part of @b-naber's comment, maybe I'm missing something?

icedrocket commented:
The behavior you expect seems to be a good way to go, but a few questions remain.

- Isn't `-Cdebuginfo` an option to set the level of debug information from rustc? Even if `-Cdebuginfo=none`, there may be other debug information included, and I'm not sure if it's desirable to ignore it.

- [The documentation](https://github.com/rust-lang/rust/blob/3f39cae1199a2a0217c3646a16d1ae7fa599130b/src/doc/rustc/src/codegen-options/index.md?plain=1#L542-L545) already mentions that `-Csplit-debuginfo=off` is supported on all platforms except MSVC, but that's not always the case in practice. AFAIK, the default value for [`split_debuginfo`](https://github.com/rust-lang/rust/blob/3f39cae1199a2a0217c3646a16d1ae7fa599130b/compiler/rustc_target/src/spec/mod.rs#L2169) in `TargetOptions` is `Off`, and if we decide to completely prohibit unsupported choices, we'll have to make them optional, like `Option<SplitDebuginfo>`.

icedrocket commented:
> > `off` is not supported on MSVC, so there wouldn't be a way to prevent a PDB file from being created if we ignore strip afaict.
> 
> Reading this part of @b-naber's comment, maybe I'm missing something?

Initially, the pull request was just to change the `strip` so that it only applied to binaries. The concern was that the change would always pass `/DEBUG:FULL` to MSVC, which would cause the PDB file to always be created.

icedrocket commented:
> > `off` is not supported on MSVC, so there wouldn't be a way to prevent a PDB file from being created if we ignore strip afaict.
> 
> Reading this part of @b-naber's comment, maybe I'm missing something?

Initially, the pull request was just to change `-Cstrip` so that it only applies to binaries. The concern at the time was that the change would always pass `/DEBUG:FULL` to MSVC, so the PDB file would always be created.

icedrocket commented:
> > `off` is not supported on MSVC, so there wouldn't be a way to prevent a PDB file from being created if we ignore strip afaict.
> 
> Reading this part of @b-naber's comment, maybe I'm missing something?

Initially, the pull request was just to change `-Cstrip` so that it only applies to binaries. The concern at the time was that the change would always pass `/DEBUG:FULL` to MSVC, so there would be no way to prevent a PDB file from being created.

icedrocket commented:
The behavior you expect seems to be a good way to go, but I have a few concerns about it.

- Isn't `-Cdebuginfo` an option to set the level of debug information from rustc? Even if `-Cdebuginfo=none`, there may be other debug information included, and I'm not sure if it's desirable to ignore it. Two commits seem to have affected the behavior: [one](https://github.com/rust-lang/rust/commit/a6c8c7d0db7f4d13960b0270419285d3e1fbf5ac) to always generate a PDB file, and [one](https://github.com/rust-lang/rust/commit/a6c2f73b6e0d24af5396355886b26bb23885c37e) to add `-Zstrip` option.

- [The documentation](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/src/doc/rustc/src/codegen-options/index.md?plain=1#L542-L545) already mentions that `-Csplit-debuginfo=off` is supported on all platforms except MSVC, but that's not always the case in practice. If `off` is a value that can be ignored if it is not supported, it is inconsistent to disallow it only on MSVC. I found [a comment](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/compiler/rustc_target/src/spec/mod.rs#L2167-L2168) that mentions that `SplitDwarfKind::None`, which was used before `-Csplit-debuginfo` was stabilized, has a target-specific meaning.

> > `off` is not supported on MSVC, so there wouldn't be a way to prevent a PDB file from being created if we ignore strip afaict.
> 
> Reading this part of @b-naber's comment, maybe I'm missing something?

Initially, the pull request was just to change `-Cstrip` so that it only applies to the executable. The concern at the time was that the change would always pass `/DEBUG:FULL` to MSVC, so there would be no way to prevent a PDB file from being created.

davidtwco commented:
> * Isn't `-Cdebuginfo` an option to set the level of debug information from rustc? Even if `-Cdebuginfo=none`, there may be other debug information included, and I'm not sure if it's desirable to ignore it. Two commits seem to have affected the behavior: [one](https://github.com/rust-lang/rust/commit/a6c8c7d0db7f4d13960b0270419285d3e1fbf5ac) to always generate a PDB file, and [one](https://github.com/rust-lang/rust/commit/a6c2f73b6e0d24af5396355886b26bb23885c37e) to add `-Zstrip` option.

I think the behaviour from the first of those commits is correct. It seems like we always produce a `pdb` file on Windows, which is necessary for backtraces (if those old issues are still correct), whether that contains lots of debuginfo or not is controlled by `-g`/`-Cdebuginfo`.

> * [The documentation](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/src/doc/rustc/src/codegen-options/index.md?plain=1#L542-L545) already mentions that `-Csplit-debuginfo=off` is supported on all platforms except MSVC, but that's not always the case in practice.  I found [a comment](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/compiler/rustc_target/src/spec/mod.rs#L2167-L2168) that mentions that `SplitDwarfKind::None`, which was used before `-Csplit-debuginfo` was stabilized, has a target-specific meaning.

I believe it is always the case in practice. rustc shouldn't accept `-Csplit-debuginfo=off` on MSVC because the compiler has no way of disabling split debuginfo on this platform, it's always `packed` because MSVC puts debuginfo in a `pdb` file. I probably wrote that comment in `rustc_target`, I don't think it's particularly relevant.

---

@wesleywiser, do you have any thoughts here? It seems like the equivalent of stripping on Windows platforms is to remove `pdb` files, but controlling that with `-Cstrip` is quite confusing.

It seems to me that there are two options here:

- Stop `-Cstrip` from influencing pdb file generation. Users would no longer have a way of disabling pdb generation entirely, since `-Cdebuginfo=none` would still generate a small `pdb` file. Configurations like that in #114215 would produce debuginfo on all platforms, which is what I'd expect.
- Keep the current behaviour. Configurations like that in #114215 are confusing, but `strip = "symbols"` on Windows means removing the `pdb` file entirely, as that's the only way to achieve that outcome.

Neither are ideal, but absent a `/STRIP` option we can pass to the MSVC linker, we seem a bit stuck (there is [`/PDBSTRIPPED`](https://learn.microsoft.com/en-us/cpp/build/reference/pdbstripped-strip-private-symbols?view=msvc-170), but doesn't seem to work like the equivalent `ld` options).

icedrocket commented:
> > * [The documentation](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/src/doc/rustc/src/codegen-options/index.md?plain=1#L542-L545) already mentions that `-Csplit-debuginfo=off` is supported on all platforms except MSVC, but that's not always the case in practice.  I found [a comment](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/compiler/rustc_target/src/spec/mod.rs#L2167-L2168) that mentions that `SplitDwarfKind::None`, which was used before `-Csplit-debuginfo` was stabilized, has a target-specific meaning.
> 
> I believe it is always the case in practice. rustc shouldn't accept `-Csplit-debuginfo=off` on MSVC because the compiler has no way of disabling split debuginfo on this platform, it's always `packed` because MSVC puts debuginfo in a `pdb` file. I probably wrote that comment in `rustc_target`, I don't think it's particularly relevant.

Sorry, I've confused you. I meant that targets where debuginfo is not supported, split debuginfo is of course not supported either.

In fact, what I was suggesting was to change the current definition of `-Csplit-debuginfo=off`: use split debuginfo if possible, but if not, try other possible alternatives. With that definition, setting `-Cstrip=symbols` would prevent the creation of a `*.pdb` file, which is an alternative to debuginfo in the executable.

> * Stop `-Cstrip` from influencing pdb file generation. Users would no longer have a way of disabling pdb generation entirely, since `-Cdebuginfo=none` would still generate a small `pdb` file. Configurations like that in [Inconsistent behavior across platform of split-debuginfo #114215](https://github.com/rust-lang/rust/issues/114215) would produce debuginfo on all platforms, which is what I'd expect.

Maybe rustc should check to see if a user-set `/DEBUG` flag exists, and if not, set `/DEBUG:FULL` by default?

icedrocket commented:
> > * [The documentation](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/src/doc/rustc/src/codegen-options/index.md?plain=1#L542-L545) already mentions that `-Csplit-debuginfo=off` is supported on all platforms except MSVC, but that's not always the case in practice.  I found [a comment](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/compiler/rustc_target/src/spec/mod.rs#L2167-L2168) that mentions that `SplitDwarfKind::None`, which was used before `-Csplit-debuginfo` was stabilized, has a target-specific meaning.
> 
> I believe it is always the case in practice. rustc shouldn't accept `-Csplit-debuginfo=off` on MSVC because the compiler has no way of disabling split debuginfo on this platform, it's always `packed` because MSVC puts debuginfo in a `pdb` file. I probably wrote that comment in `rustc_target`, I don't think it's particularly relevant.

Sorry, I've caused the confusion here. I meant that targets where debuginfo is not supported, split debuginfo is of course not supported either.

In fact, what I was suggesting was to change the current definition of `-Csplit-debuginfo=off`: use debuginfo in the executable if possible, but if not, try other possible alternatives. With that definition, setting `-Cstrip=symbols` would prevent the creation of a `*.pdb` file, which is an alternative to debuginfo in the executable.

> * Stop `-Cstrip` from influencing pdb file generation. Users would no longer have a way of disabling pdb generation entirely, since `-Cdebuginfo=none` would still generate a small `pdb` file. Configurations like that in [Inconsistent behavior across platform of split-debuginfo #114215](https://github.com/rust-lang/rust/issues/114215) would produce debuginfo on all platforms, which is what I'd expect.

Maybe rustc should check to see if a user-set `/DEBUG` flag exists, and if not, set `/DEBUG:FULL` by default?

icedrocket commented:
> > * [The documentation](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/src/doc/rustc/src/codegen-options/index.md?plain=1#L542-L545) already mentions that `-Csplit-debuginfo=off` is supported on all platforms except MSVC, but that's not always the case in practice.  I found [a comment](https://github.com/rust-lang/rust/blob/3ad8e2d129a67b0dc450d6ef1100ae41d2cd4b8a/compiler/rustc_target/src/spec/mod.rs#L2167-L2168) that mentions that `SplitDwarfKind::None`, which was used before `-Csplit-debuginfo` was stabilized, has a target-specific meaning.
> 
> I believe it is always the case in practice. rustc shouldn't accept `-Csplit-debuginfo=off` on MSVC because the compiler has no way of disabling split debuginfo on this platform, it's always `packed` because MSVC puts debuginfo in a `pdb` file. I probably wrote that comment in `rustc_target`, I don't think it's particularly relevant.

Sorry, I've caused the confusion here. I meant that targets where debuginfo is not supported, `-Csplit-debuginfo=off` is of course not supported either.

In fact, what I was suggesting was to change the current definition of `-Csplit-debuginfo=off`: use embedding debuginfo in the executable if possible, but if not, try other possible alternatives. With that definition, setting `-Cstrip=symbols` would prevent the creation of a `*.pdb` file, which is an alternative to embedding debuginfo in the executable.

> * Stop `-Cstrip` from influencing pdb file generation. Users would no longer have a way of disabling pdb generation entirely, since `-Cdebuginfo=none` would still generate a small `pdb` file. Configurations like that in [Inconsistent behavior across platform of split-debuginfo #114215](https://github.com/rust-lang/rust/issues/114215) would produce debuginfo on all platforms, which is what I'd expect.

Maybe rustc should check to see if a user-set `/DEBUG` flag exists, and if not, set `/DEBUG:FULL` by default?

icedrocket commented:
@davidtwco I agree that there doesn't seem to be an ideal solution. If there's no good way to improve the issue, I'm going to close this pull request and let others contribute in the future, but I have some unresolved thoughts.

The only clear benefit to ignoring the `-C strip` is that it might be less confusing, so it might be better to keep the current behavior. Since compilation flags can only be set per profile or per target in the config file, `-C strip` will only be able to be set per target when used cross-platform, but that's not a big deal since those can also be set using the build script.

The second thing I was trying to do with the pull request change is to emulate the behavior on the ELF platform as much as possible on all platforms, even if imperfectly. That's why I mentioned "fallback behavior" and I know this might seem a bit odd, but from a user perspective, I thought it might be beneficial because the behavior is more consistent across all platforms.

Older contributors to the conversation here have expressed the opinion that MSVC was not originally designed to include debug information in the executable, so making `-Csplit-debuginfo=off` available for MSVC might not be in line with the original design intent.

I'm not sure what Rust's design philosophy or direction is, but if it is to reduce unclear expressions, I think what I've tried to do may not be in line with that.

What are your thoughts on this?

icedrocket commented:
@davidtwco I agree that there doesn't seem to be an ideal solution. If I don't see a good way to improve the issue, I'm going to close this pull request and allow others to contribute in the future. However, I have a few unanswered questions and thoughts.

The only clear benefit to ignoring the `-C strip` is that it might be less confusing, so it might be better to keep the current behavior. Since compilation flags can only be set per profile or per target in the config file, `-C strip` will only be able to be set per target when used cross-platform, but that's not a big deal since those can also be set using the build script.

The second thing I was trying to do with the pull request change is to emulate the behavior on the ELF platform as much as possible on all platforms, even if imperfectly. That's why I mentioned "fallback behavior" and I know this might seem a bit odd, but from a user perspective, I thought it might be beneficial because the behavior is more consistent across all platforms.

Older contributors to the conversation here have expressed the opinion that MSVC was not originally designed to include debug information in the executable, so making `-Csplit-debuginfo=off` available for MSVC might not be in line with the original design intent.

I'm not sure what Rust's design philosophy or direction is, but if it is to reduce unclear expressions, I think what I've tried to do may not be in line with that.

What are your thoughts on this?

icedrocket commented:
@davidtwco I agree that there doesn't seem to be an ideal solution. If I couldn't find a good way to improve the issue, I'm going to close this pull request and allow others to contribute in the future. However, I have a few unanswered questions and thoughts.

The only clear benefit to ignoring the `-C strip` is that it might be less confusing, so it might be better to keep the current behavior. Since compilation flags can only be set per profile or per target in the config file, `-C strip` will only be able to be set per target when used cross-platform, but that's not a big deal since those can also be set using the build script.

The second thing I was trying to do with the pull request change is to emulate the behavior on the ELF platform as much as possible on all platforms, even if imperfectly. That's why I mentioned "fallback behavior" and I know this might seem a bit odd, but from a user perspective, I thought it might be beneficial because the behavior is more consistent across all platforms.

Older contributors to the conversation here have expressed the opinion that MSVC was not originally designed to include debug information in the executable, so making `-Csplit-debuginfo=off` available for MSVC might not be in line with the original design intent.

I'm not sure what Rust's design philosophy or direction is, but if it is to reduce unclear expressions, I think what I've tried to do may not be in line with that.

What are your thoughts on this?

davidtwco commented:
I'll nominate this for discussion in a team meeting.

apiraino commented:
pull request was discussed in the T-compiler triage meeting (on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202024-02-15/near/421683922)).

@_wesleywiser volunteered to investigate and report (thanks!)

@rustbot -I-compiled-nominated

michaelwoerister commented:
As a data point: for the trim-paths RFC, it looks like Cargo will compile binaries on macOS with `-Cstrip=debuginfo -Csplit-debuginfo=packed` for the `object` trim-level where the binary should be sanitized but debuginfo is left unsanitized. So on that platform, we are going to rely on the combination of `-Cstrip` and `-Csplit-debuginfo` to strip the binary but leave debuginfo intact.

workingjubilee commented:
Without this change, and with a combination of other changes, we have [simply broken backtraces](https://github.com/rust-lang/rust/issues/122857) on Windows. whoops lmao

icedrocket commented:
See also rust-lang/compiler-team#721. One option would be to implement and stabilize that MCP and then change `-Cstrip` to be ignored on MSVC. This would probably require a separate pull request.

michaelwoerister commented:
I think we should decouple `-Cstrip` from PDB file generation, and define the meaning of `-Csplit-debuginfo=off` to be equivalent to passing `/DEBUG:NONE` to the linker for MSVC. I don't see a reason to tie PDB (or other separate debuginfo) behavior to `-Cstrip`. `-Csplit-debuginfo=off` is currently not supported on MSVC, so we are free to give it a meaningful behavior. It would still not be the default and the documentation would state that it leads to faster link times but breaks backtraces. 

If the `-Csplit-debuginfo=off` part is controversial, I propose that we still decouple `-Cstrip` from PDB file generation. `-Cstrip` would be ignored on MSVC.

cc @rust-lang/wg-debugging

michaelwoerister commented:
Btw, I would be happy to put together a self-contained proposal so that @icedrocket can update the PR and we can start an FCP. But I'd like some feedback from @davidtwco about [this comment](https://github.com/rust-lang/rust/pull/115120#issuecomment-1875265139):

> rustc shouldn't accept -Csplit-debuginfo=off on MSVC because the compiler has no way of disabling split debuginfo on this platform, it's always packed because MSVC puts debuginfo in a pdb file.

The behavior of `-Csplit-debuginfo=off` on MSVC seems like the only open question, we'd need to answer before doing an FCP.

wesleywiser commented:
> define the meaning of -Csplit-debuginfo=off to be equivalent to passing /DEBUG:NONE to the linker for MSVC

I think ideally `-Csplit-debuginfo` would only control _where_ debuginfo ends up after compilation completes while `-Cdebuginfo-level` would control what debuginfo (if any) is generated. As such, `-Csplit-debuginfo=off` probably doesn't make sense on `*-windows-msvc`? 

> I propose that we still decouple -Cstrip from PDB file generation. -Cstrip would be ignored on MSVC.

This makes sense to me. `-Cstrip` in general seems weird to me as a Windows developer because the symbol table is so much sparser and we already have split debuginfo by default. I think ignoring the `-Cstrip` flag for `*-windows-msvc` is reasonable.

davidtwco commented:
> > define the meaning of -Csplit-debuginfo=off to be equivalent to passing /DEBUG:NONE to the linker for MSVC
> 
> 
> 
> I think ideally `-Csplit-debuginfo` would only control _where_ debuginfo ends up after compilation completes while `-Cdebuginfo-level` would control what debuginfo (if any) is generated. As such, `-Csplit-debuginfo=off` probably doesn't make sense on `*-windows-msvc`? 
> 
> 
> 
> > I propose that we still decouple -Cstrip from PDB file generation. -Cstrip would be ignored on MSVC.
> 
> 
> 
> This makes sense to me. `-Cstrip` in general seems weird to me as a Windows developer because the symbol table is so much sparser and we already have split debuginfo by default. I think ignoring the `-Cstrip` flag for `*-windows-msvc` is reasonable.

I'm literally sitting next to Wesley and I co-sign this conclusion.

michaelwoerister commented:
> I think ideally -Csplit-debuginfo would only control where debuginfo ends up

Well, the way I've been interpreting this is that `-Csplit-debuginfo=off` means that debuginfo goes into the debuginfo section(s) of the binary -- which for MSVC just happens to be `/dev/null` 😉

However, I'm happy if we can fix the `-Cstrip` part. We can revisit the `/DEBUG:NONE` part at some point the future.

bjorn3 commented:
What about only allowing `-Csplit-debuginfo=off` on MSVC when also using `-Cstrip=symbols`? And if in the future MSVC will allow embedding debuginfo in the executable itself for whatever reason, then allow `-Csplit-debuginfo=off` without `-Cstrip=symbols` to put the debuginfo in the executable. For `-Csplit-debuginfo=unpacked/packed` stripping doesn't make sense to me as stripping is supposed to affect the executable itself and not separate debuginfo.

workingjubilee commented:
/Z7 isn't relevant to this discussion, right? https://learn.microsoft.com/en-us/cpp/build/reference/z7-zi-zi-debug-information-format?view=msvc-170

michaelwoerister commented:
> /Z7 isn't relevant to this discussion, right? https://learn.microsoft.com/en-us/cpp/build/reference/z7-zi-zi-debug-information-format?view=msvc-170

Nope, that is only for object files, not the final EXE and DLL files. 
This is a pretty good summary of debuginfo on MSVC: https://learn.microsoft.com/en-us/cpp/build/reference/debug-generate-debug-info?view=msvc-170#remarks

michaelwoerister commented:
## PROPOSAL

tl;dr - Define `-Cstrip` to only ever affect the binary; no other build artifacts.

This is the diff to the documentation of a the `-Cstrip` flag (additions in bold):

> ## strip
> 
> The option `-C strip=val` controls stripping of debuginfo and similar auxiliary
> data from binaries during linking.
> 
> Supported values for this option are:
> 
> - `none` - debuginfo and symbols ~~(if they exist) are copied to the produced
>   binary or separate files depending on the target (e.g. `.pdb` files in case
>   of MSVC).~~ **are not modified.**
> - `debuginfo` - debuginfo sections and debuginfo symbols from the symbol table
>   section are stripped at link time. ~~and are not copied to the produced binary
>   or separate files.~~ **This has no effect on debuginfo outside the binary, like `.pdb` or `.dbg` files.**
> - `symbols` - same as `debuginfo`, but the rest of the symbol table section is
>   stripped as well if the linker supports it.

Remarks:
- A consequence of this is that there is no way to tell `rustc` to not generate a PDB file. We might revisit the question later because it could be beneficial for build times, but commonly a PDB file is needed even for optimized, `-Cdebuginfo=0` builds if backtraces are required.
- The behavior differs from the [MCP](https://github.com/rust-lang/compiler-team/issues/669) in that `-Csplit-debuginfo=off` does not prevent the compiler from generating a PDB. However, the behavior does make it possible to have consistency between platforms. Cargo can stop [special-casing MSVC targets](https://github.com/rust-lang/cargo/pull/13630).
- This will change the behavior of a stable copmpiler option but it should not break anything. The only effect is that the compiler will generate a PDB file where it didn't before.

@rfcbot fcp merge

retep998 commented:
I think there's a general process failure in that people discussing the hows of implementation without ever discussing the **why**.

**Why** do people strip debug symbols? To make the binary size smaller.

Does preventing the PDB file from being generated make the binary size smaller? No, it does not. If you want to package up everything to ship to end users, just *don't* ship the PDB file.

So then **why** would people not want to generate PDB files?
Does it save on linker time? I don't see any comparisons of linker time here, but in my experience the time saved is negligible.
Does it break backtraces? Yes, Windows MSVC relies on the PDB to generate useful backtraces, and thus *any* consideration to remove the PDB file needs to weigh the downside of losing out on backtraces.

Given the significant downsides of not having a PDB file, without having the primary benefit of reduced binary size...
I consider it perfectly reasonable to *ignore* the `strip` option on Windows MSVC, or even warn the user that the option makes no sense as there is nothing to be stripped.

As for the `split-debuginfo` option, that simply does not even make sense to ask as a question on Windows MSVC. Debuginfo is always in a PDB file, so there's no point in trying to interpret the user's intent here. Most of the time this would only ever be specified on Windows by users who are working on other platforms and specify the option for all platforms without even considering what makes sense on Windows.

Thus, I consider it perfectly reasonable to *ignore* the `split-debuginfo` option entirely on Windows MSVC, or even warn the user that hey this option makes no sense on this platform.


Therefore my answer is both options should just do nothing or warn on Windows MSVC.

workingjubilee commented:
I think we should not warn on Windows when strip is passed, so people can have a consistent cross-platform experience. People generally prefer to silence warnings rather than ignore them, in Rust, and it is silly to have to disable an option that is going to be ignored anyways.

I agree that there is an issue here with confusing "how" and "why". My impression of what that means, however, is that I think if we wanted to have a truly coherent story about what debuginfo was generated, we'd reconsider the entire premise of all these flags, and go back to basics, starting with contemplating `-Cdebuginfo`. I think that is *wildly* out of scope of this PR and this FCP, however, but I can elaborate on such a proposal elsewhere if people express interest.

bjorn3 commented:
> As for the split-debuginfo option, that simply does not even make sense to ask as a question on Windows MSVC.

As I understand it on MSVC there is still the option to only emit just enough debuginfo in the pdb file to lookup the actual debuginfo in the original object files. This would correspond to `-Csplit-debuginfo=unpacked`.

icedrocket commented:
My intention with this pull request was simply to reduce inconsistent behavior. I wasn't sure if removing the ability to disable PDB file generation was desirable, which is why I suggested a possible alternative using `-Csplit-debuginfo=off` (I've seen many platforms default to `off` regardless of whether it's supported or not).

As long as useful backtrace relies on PDB files, I think it's appropriate to ignore both options as @retep998 mentioned.

michaelwoerister commented:
Thanks for updating the PR, @icedrocket!

For the record, I don't think any of the new comments are in conflict with the FCP (which only proposes to ignore `-Cstrip` when targeting MSVC). Please let us know if there are any concerns with the FCP or the current version (https://github.com/rust-lang/rust/pull/115120/commits/95adb6d7b520230971522c8806cc6ab518e419c5) of the PR.

bors commented:
:umbrella: The latest upstream changes (presumably #123336) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Commit:
Ignore `-C strip` on MSVC

michaelwoerister commented:
Thanks for keeping at this, @icedrocket!

@bors r+ rollup=never

bors commented:
:pushpin: Commit e82f46ab7226b69b29d5265c94647eeb5ca1e278 has been approved by `michaelwoerister`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=michaelwoerister e82f46ab7226b69b29d5265c94647eeb5ca1e278 -->
<!-- homu: {"type":"Approved","sha":"e82f46ab7226b69b29d5265c94647eeb5ca1e278","approver":"michaelwoerister","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit e82f46ab7226b69b29d5265c94647eeb5ca1e278 with merge 7f2fc33da6633f5a764ddc263c769b6b2873d167...
<!-- homu: {"type":"BuildStarted","head_sha":"e82f46ab7226b69b29d5265c94647eeb5ca1e278","merge_sha":"7f2fc33da6633f5a764ddc263c769b6b2873d167"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8783694140/job/24106364114)
Approved by: michaelwoerister
Pushing 7f2fc33da6633f5a764ddc263c769b6b2873d167 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"michaelwoerister","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8783694140/job/24106364114"},"merge_sha":"7f2fc33da6633f5a764ddc263c769b6b2873d167"} -->

rust-timer commented:
Finished benchmarking commit (7f2fc33da6633f5a764ddc263c769b6b2873d167): [comparison URL](https://perf.rust-lang.org/compare.html?start=290d792411566c94273e5d1a980a1ab91f1d950b&end=7f2fc33da6633f5a764ddc263c769b6b2873d167&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=290d792411566c94273e5d1a980a1ab91f1d950b&end=7f2fc33da6633f5a764ddc263c769b6b2873d167&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=290d792411566c94273e5d1a980a1ab91f1d950b&end=7f2fc33da6633f5a764ddc263c769b6b2873d167&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4% | [2.4%, 2.4%] | 1     |
| Regressions ❌ <br /> (secondary)  | -    | -            | 0     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 2.4% | [2.4%, 2.4%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=290d792411566c94273e5d1a980a1ab91f1d950b&end=7f2fc33da6633f5a764ddc263c769b6b2873d167&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=290d792411566c94273e5d1a980a1ab91f1d950b&end=7f2fc33da6633f5a764ddc263c769b6b2873d167&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 671.419s -> 670.936s (-0.07%)
**Artifact size**: 315.41 MiB -> 315.45 MiB (0.01%)

weihanglo commented:
Since this is in 1.79.0 already. Should we revert [this PR](https://github.com/rust-lang/cargo/pull/13630) in Cargo?

evmar commented:
I found this change when I was trying to figure out why my link lines always include /DEBUG.

I might be misreading the code, but it seems .debuginfo() is called unconditionally, which means that flag is now passed unconditionally, regardless of any profile's debug or strip settings(?).  This not only produces a pdb, it also produces a debug info section in the resulting binary.  It is the sort of thing I would expect some sort of debug-stripping flag to remove.

Note how most of these sections are empty, but the debug directory one isn't.

```
$ llvm-objdump -x target/i586-pc-windows-msvc/release/rdll.dll 
[...]
The Data Directory
Entry 0 00002059 00000041 Export Directory [.edata (or where ever we found it)]
Entry 1 00000000 00000000 Import Directory [parts of .idata]
Entry 2 00000000 00000000 Resource Directory [.rsrc]
Entry 3 00000000 00000000 Exception Directory [.pdata]
Entry 4 00000000 00000000 Security Directory
Entry 5 00000000 00000000 Base Relocation Directory [.reloc]
Entry 6 00002000 00000038 Debug Directory
[...]
```

That data visualized in ghidra is

<img width="647" alt="image" src="https://github.com/user-attachments/assets/2e6c58bf-c9db-40e3-b779-fd5acf3911ef">

workingjubilee commented:
@evmar Hi, can you open a new issue about your problem? Thanks!

retep998 commented:
> I found this change when I was trying to figure out why my link lines always include /DEBUG.
> 
> I might be misreading the code, but it seems .debuginfo() is called unconditionally, which means that flag is now passed unconditionally, regardless of any profile's debug or strip settings(?). This not only produces a pdb, it also produces a debug info section in the resulting binary. It is the sort of thing I would expect some sort of debug-stripping flag to remove.

`/DEBUG` being unconditionally applied was always intentional for several reasons:

- The cost of `/DEBUG` being enabled is low, as the debug info section is only a few bytes. The link time penalty is also fairly low.
- It allows backtraces on Windows to actually have symbols instead of just addresses. On other platforms backtraces have symbols even without debuginfo due to the nature of how they handle linking. Being able to have this on Windows as well is a huge benefit.
- For the rare occasion where someone *really* insists on not having a PDB generated, nor a debug info section, they can pass a custom linker argument of `/DEBUG:NONE` to override `/DEBUG`.

evmar commented:
(@retep998 Happy to open a new bug if it helps, lemme know)

/DEBUG:NONE is a good idea!

I tried adding to my build.rs
```
    println!("cargo::rustc-link-arg=/DEBUG:NONE");
```

And then built with
```
$ RUSTFLAGS='--print link-args' cargo build --release --target i586-pc-windows-msvc  --examples      
```

And the tail of the printed link line looks like:
```
[...] "/DEBUG" "/PDBALTPATH:%_PDB%" "/Brepro" "/noentry" "/nodefaultlib" "/DEBUG:NONE"
```

(where /DEBUG is from this PR, and noentry etc. are other flags I have already added) but the resulting binary seems to still include this section.

I think the relevant flag-parsing bit of lld is [here](https://github.com/llvm/llvm-project/blob/0f723eb67197421caf6504a7e4594751040b1924/lld/COFF/Driver.cpp#L1693) and it seems like that ought to be enough, so now I am just confused.

Is there a way I can rerun the command directly to see if removing /DEBUG helps?  It seems it refers to temporary files so if I copy-paste it to try it myself it fails.

evmar commented:
I think I found it, it's [this](https://github.com/llvm/llvm-project/blob/0f723eb67197421caf6504a7e4594751040b1924/lld/COFF/Writer.cpp#L1163-L1183) logic, which appears to have been tickled in my case by the /Brepro flag.  Not your problem.

I am happy with the /DEBUG:NONE solution, I agree with your take that this is sufficiently esoteric that it doesn't need an easy user-facing knob.

evmar commented:
Just in case anyone finds this in the future, I asked the author of the relevant bits of LLVM and he pointed me at some relevant discussion:
https://reviews.llvm.org/D51652

tldr LLVM is making a subtle and intentional choice to include this debug section.
